scriptoocloud/src/main/webapp/WEB-INF/web.xml

```diff
@@ -66,13 +82,21 @@
                 org.springframework.web.context.support.AnnotationConfigWebApplicationContext
             </param-value>
         </init-param>
-        <!-- Again, config locations must consist of one or more comma- or space-delimited 
+        <!-- Again, config locations must consist of one or more comma- or space-delimited
             and fully-qualified @Configuration classes -->
         <init-param>
             <param-name>contextConfigLocation</param-name>
             <param-value>edu.asu.diging.scriptoocloud.config.MvcConfig</param-value>
         </init-param>
         <load-on-startup>1</load-on-startup>
+
+        <!-- Added to support file uploads -->
+        <multipart-config>
+            <location>/tmp</location>
```

Comment:
this is an absolute path, right? Tomcat won't have permission to put anything at root. This would need to be a configurable property.

Reply:
I also don't think that you need that. You can use `commons-fileupload` and declare a `CommonsMultipartResolver` bean instead.

Reply:
https://docs.spring.io/spring-framework/docs/current/reference/html/web.html#mvc-multipart
https://www.baeldung.com/spring-file-upload

Reply:
I removed the `MultipartResolver` configuration from `web.xml` and put in Java configuration as requested.  Initially, I then received an Access Denied error when uploading a `DataFile`.  According to Spring Security documentation, there are two ways to handle this (each with their own tradeoffs):

[https://docs.spring.io/spring-security/site/docs/4.0.0.RC1/reference/htmlsingle/#csrf-multipart](https://docs.spring.io/spring-security/site/docs/4.0.0.RC1/reference/htmlsingle/#csrf-multipart)

I included the CSRF as query parameters as below:
```
<form method="POST"
    class="file-upload-form"
    th:action="@{/datasets/{id}/upload?{name}={token}(id=${dataset.id}, name=${_csrf.parameterName}, token=${_csrf.token})}"
    enctype="multipart/form-data">
    ```
This solution is working, but I was wondering if the first solution in the documentation is preferred - that is, should I override `beforeSpringSecurityFilterChain`?  If so, should I create a new config class: `SecurityApplicationInitializer` as in the documentation?

Reply:
No, this is fine. If you override the filter then the file is uploaded before the security checks (so it'll also be uploaded even if the user is not authorized). I prefer it this way even though the token is passed in the url.